apps/main/src/llamalend/mutations/useLlammaMutation.ts

```diff
@@ -0,0 +1,130 @@
+ *
+ * @remarks
+ * - The error string content is not standardized and does not have a guaranteed form
+ * - Making errors prettier is considered out of scope
+ * - Successfully determining if an error was simply a user cancelling a transaction is out of scope
+ * - User rejection errors (containing "User rejected the request") are ignored and do not throw
+ */
+function throwIfError(data: unknown) {
+  // If the data contains an error object, it probably means the transaction failed even though nothing
+  // before that was going wrong. In other words, 'failed successfully'.
+  if (data != null && typeof data === 'object' && 'error' in data) {
+    // Not fail proof, as the content of the error string is not standardized
+    // and does not have a guaranteed form. Making errors prettier is out of scope and succesfully
+    // determined if it was simple a user cancelling a transaction is out of scope as well.
+    if (typeof data.error === 'string' && !data.error.includes('User rejected the request')) {
+      throw new Error(data.error)
+    }
+  }
+}
+
+/** Context created in onMutate to all callbacks other than mutationFn that also validates */
+type Context = {
+  provider: NonNullable<ReturnType<typeof useWallet>['provider']>
+  curve: NonNullable<ReturnType<typeof useConnection>['llamaApi']>
+  market: LlamaMarketTemplate
+  pendingNotification: ReturnType<typeof notify>
+}
+
+/**
+ * Custom context specially for the mutation function.
+ * Usually mutations don't have a context, but they do
+ * when you use our specialized llamma mutation.
+ */
+type MutationFnContext = {
+  market: LlamaMarketTemplate
+}
+
+type LlammaMutationOptions<TVariables, TData> = {
+  /** The llamma market id */
+  marketId: string | null | undefined
+  mutationKey: readonly unknown[]
+  /** Function that performs the mutation operation */
+  mutationFn: (variables: TVariables, context: MutationFnContext) => Promise<TData>
+  /** Message to display during pending state */
+  pendingMessage: string | ((variables: TVariables, market: LlamaMarketTemplate) => string)
+  /** Callback executed on successful mutation */
+  onSuccess?: (data: TData, variables: TVariables, context: Context) => void
+  /** Callback executed on mutation error */
+  onError?: (error: Error, variables: TVariables, context: Context) => void
+  /** Callback executed after mutation settles (success or error) */
+  onSettled?: (
+    data: TData | undefined,
+    error: Error | null,
+    variables: TVariables,
+    context: Context | undefined,
+  ) => void
+}
+
+/**
+ * Custom hook for handling llamma-related mutations with automatic wallet and API validation
+ * Could argue for a refactor to validate with vest like we do for queries, but I'd rather keep
+ * it simple for now. Maybe another time, for now we're just doing a quick llamma specialization
+ * with simple throwing errors.
+ */
+export function useLlammaMutation<TVariables, TData>({
```

Comment:
Maybe this generic mutation is worth it, maybe not, couldn't really decide, but it's here now. The upside of this hook really is just validation, easy access to a market object and automatic notification management. If we decide not to use this hook we might forget to handle notifications here and there.